server/datastore/mysql/migrations/tables/20240228111134_AddScriptContentsTableAndRelationships.go

```diff
@@ -0,0 +1,190 @@
+	savedScriptsLookup := make(map[string][]uint, len(savedScripts))
+	for _, s := range hostScripts {
+		hexChecksum := md5ChecksumScriptContent(s.ScriptContents)
+		hostScriptsLookup[hexChecksum] = append(hostScriptsLookup[hexChecksum], s.ID)
+		if _, err := txx.Exec(insertScriptContentsStmt, hexChecksum, s.ScriptContents); err != nil {
+			return fmt.Errorf("create script_contents from host_script_results: %w", err)
+		}
+	}
+	for _, s := range savedScripts {
+		hexChecksum := md5ChecksumScriptContent(s.ScriptContents)
+		savedScriptsLookup[hexChecksum] = append(savedScriptsLookup[hexChecksum], s.ID)
+		if _, err := txx.Exec(insertScriptContentsStmt, hexChecksum, s.ScriptContents); err != nil {
+			return fmt.Errorf("create script_contents from saved scripts: %w", err)
+		}
+	}
+
+	alterAddHostScriptsStmt := `
+ALTER TABLE host_script_results
+	ADD COLUMN script_content_id INT(10) UNSIGNED NULL,
+	ADD FOREIGN KEY (script_content_id) REFERENCES script_contents (id) ON DELETE CASCADE`
+	if _, err := tx.Exec(alterAddHostScriptsStmt); err != nil {
+		return fmt.Errorf("alter add column to table host_script_results: %w", err)
+	}
+
+	alterAddScriptsStmt := `
+ALTER TABLE scripts
+	ADD COLUMN script_content_id INT(10) UNSIGNED NULL,
+	ADD FOREIGN KEY (script_content_id) REFERENCES script_contents (id) ON DELETE CASCADE`
+	if _, err := tx.Exec(alterAddScriptsStmt); err != nil {
+		return fmt.Errorf("alter add column to table scripts: %w", err)
+	}
+
+	updateHostScriptsStmt := `
+UPDATE
+	host_script_results
+SET
+	script_content_id = (SELECT id FROM script_contents WHERE md5_checksum = UNHEX(?)),
+	updated_at = updated_at
+WHERE
+	id = ?`
+
+	// for saved scripts, the `updated_at` timestamp is used as the "uploaded_at"
+	// information in the UI, so we ensure that it doesn't change with this
+	// migration.
+	updateSavedScriptsStmt := `
+UPDATE
+	scripts
+SET
+	script_content_id = (SELECT id FROM script_contents WHERE md5_checksum = UNHEX(?)),
+	updated_at = updated_at
+WHERE
+	id = ?`
+
+	// insert the associated script_content_id into host_script_results and
+	// scripts
+	for hexChecksum, ids := range hostScriptsLookup {
+		for _, id := range ids {
+			if _, err := txx.Exec(updateHostScriptsStmt, hexChecksum, id); err != nil {
+				return fmt.Errorf("update host_script_results with script_content_id: %w", err)
+			}
+		}
+	}
+	for hexChecksum, ids := range savedScriptsLookup {
+		for _, id := range ids {
+			if _, err := txx.Exec(updateSavedScriptsStmt, hexChecksum, id); err != nil {
+				return fmt.Errorf("update saved scripts with script_content_id: %w", err)
+			}
+		}
+	}
+
+	// TODO(mna): we cannot drop the "script_contents" column immediately from
+	// the host_script_results and scripts tables because that would break the
+	// current code, we need to wait for the feature to be fully implemented.
+	// There's no harm in leaving it in there unused for now, as stored scripts
+	// were previously smallish.
```

Comment:
Not to be included as part of this PR, but we can add that migration at the end of the story, and we should probably also implement a `cron` job to remove any script content that isn't needed anymore? (i.e. not referenced by any saved script/host execution).